design/vela-cli/sdk_generating.md

```diff
@@ -0,0 +1,114 @@
+# SDK Generating
+
+# Introduce
+
+This proposal introduces a framework which can generate KubeVela SDK from scaffolds, templates and X-Definitions.
+
+# Background
+
+As KubeVela designs for separation of concerns, there are two roles of user: platform team and end user. They are dealing with Application in different ways.
+
+![image](../resources/separation-of-roles.png)
+
+### For platform team
+
+When platform team users assembling Application, filling in components, traits and workflow steps, they are actually assemble Application YAML. Unfortunately, due to the extensibility of KubeVela X-Definition system, the property structure of different component (also for traits, workflow steps, policies) varies. In the source code, we use a `map[string]interface{}` to accept all kinds of properties. The unstructured code make it hard examine, modify, insert values of nested properties.
+
+Before KubeVela provide the SDK, the problem is mainly solved by providing OpenAPI schema to frontend  and generate the form dynamically. There is also an enhancement scheme: ui-schema. After filling the form, the JSON returned by frontend is already satisfied the specification of X-Definition properties. This is how VelaUX works when assembling Application.
+
+### For end user
+
+Basically there are two ways for end user to use Application:
+
+1. Through dashboard provided by platform team like VelaUX. The data is kept by the internal platform. In this pattern user's behavior can be better guided with typed form and UI indications.
+2. Storing Application as part of the codebase. Then through some way (e.g. GitOps) the application is applied to cluster.
+
+### Goals & Non-Goals
+
+Goals include the following:
+1. Provide a framework to generate SDK for different languages.
+2. Provide a Go SDK as an example.
```

Comment:
```suggestion
2. Provide a Go SDK as one of the first core SDK to proof of concept.
```

Reply:
```
3. The SDK must provide functions to meet the following scenarios:
    - Discover Definition and generate the SDK API code automatically
    - Be able to do CRUD on OAM concepts including Component/Trait/Workflow/Policy. 
    - Be able to generate application YAML and directly apply to KubeVela control plane cluster.
```

Reply:
> Discover Definition and generate the SDK API code automatically

This is wrong. It should be done by sdk generator not sdk itself. SDK should not generating new API while using it. But it's fine to ask sdk generator to discover definition from Clsuter

Reply:
Yes, then add the sdk-generator in our goals.